comment/src/comment.ts

```diff
@@ -0,0 +1,280 @@
+  open: string
+  close: string
+  margin: string
+  constructor(open: string, close: string, margin: string = " ") {
+    this.open = open
+    this.close = close
+    this.margin = margin
+  }
+
+  toggle(option: CommentOption, state: EditorState): Transaction | null {
+    const selectionCommented = this.isSelectionCommented(state)
+    if (selectionCommented !== null) {
+      if (option !== CommentOption.OnlyComment) {
+        const tr = state.t()
+        const mapRef = tr.mapRef()
+        for (const range of selectionCommented) {
+          const open = range.open
+          const close = range.close
+          open.pos = mapRef.mapPos(open.pos)
+          tr.replace(open.pos - this.open.length, open.pos + open.margin, "")
+          close.pos = mapRef.mapPos(close.pos)
+          tr.replace(close.pos - close.margin, close.pos + this.close.length, "")
+        }
+
+        return tr
+      }
+    } else {
+      if (option !== CommentOption.OnlyUncomment) {
+        const tr = state.t()
+        tr.forEachRange((range: SelectionRange, tr: Transaction) => {
+          const copen = new Change(range.from, range.from, tr.startState.splitLines(this.open + this.margin))
+          const cclose = new Change(range.to, range.to, tr.startState.splitLines(this.margin + this.close))
+          tr.change([copen, cclose])
+          const shift = (this.open + this.margin).length
+          return new SelectionRange(range.anchor + shift, range.head + shift)
+        })
+
+        return tr
+      }
+    }
+
+    return null
+  }
+
+  // Determines whether all selection ranges in `state` are block-commented.
+  isSelectionCommented(state: EditorState): { open: { pos: number, margin: number }, close: { pos: number, margin: number } }[] | null {
+    let result = []
+    for (const range of state.selection.ranges) {
+      const x = this.isRangeCommented(state, range)
+      if (x === null) return null
+      result.push(x)
+    }
+    return result
+  }
+
+  // Determines if the `range` is block-commented in the given `state`.
+  // The `range` must be a valid range in `state`.
+  isRangeCommented(state: EditorState, range: SelectionRange): { open: { pos: number, margin: number }, close: { pos: number, margin: number } } | null {
+    type SearchWithType = (this: string, searchString: string, pos?: number) => boolean
+    const search = (pos: number, searchString: string, searchWith: SearchWithType, d: 1 | -1, i: 1 | 0): { pos: number, margin: number } | null => {
+      const line = state.doc.lineAt(pos)
+      const str = line.content as string
+      const ss = eatSpace(str, pos - line.start - i, d)
+      return searchWith.call(str, searchString, pos + d * ss - line.start)
+        ? { pos: pos + d * ss, margin: ss == 0 ? 0 : 1 }
+        : null
+    }
+    const startsWithOpen = search(range.from, this.open, String.prototype.endsWith, -1, 1)
+    const endsWithClose = search(range.to, this.close, String.prototype.startsWith, 1, 0)
```

Comment:
The library targets browsers (IE11) that don't have `endsWith` and `startsWith` on their strings yet, so we'll need to use a replacement for those.